File: 백준/Silver/13908. 비밀번호/비밀번호.cc
#include <iostream>
#include <algorithm>
#include <string>
using namespace std;
int n, m, a[10], ans;

bool check(int c[10])
{
	for (int i = 0; i < 10; i++)
	{
		if (a[i] && !c[i])
		{
			return false;
		}
	}
	return true;
}

void go(int len, int c[10])
{
	if (len > n)
	{
		return;
	}
	if (check(c) && len == n)
	{
		ans++;
	}

	for (int i = 0; i < 10; i++)
	{
		c[i]++;
		go(len + 1, c);
		c[i]--;
	}
}

int main()
{
	cin >> n >> m;
	for (int i = 0; i < m; i++)
	{
		int k;
		cin >> k;
		a[k] = 1;
	}
	int c[10] = {
		0,
	};
	go(0, c);
	cout << ans;
}

File: 백준/Silver/13908. 비밀번호/비밀번호.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

	static int n, m;
	static boolean[] check = new boolean[10];
	static int[] vis = new int[10];
	static int ans;

	public static void main(String[] args) throws Exception {
		var br = new BufferedReader(new InputStreamReader(System.in));

		var st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		if (m != 0) {
			st = new StringTokenizer(br.readLine());
			for (int i = 0; i < m; i++) {
				check[Integer.parseInt(st.nextToken())] = true;
			}
		}
		go(0);
		System.out.println(ans);
	}

	static void go(int len) {
		if (len == n) {
			if (isVis()) {
				ans++;
			}
			return;
		}
		for (int i = 0; i < 10; i++) {
			vis[i]++;
			go(len + 1);
			vis[i]--;
		}
	}

	static boolean isVis() {
		for (int i = 0; i < 10; i++) {
			if (check[i] && vis[i] == 0) {
				return false;
			}
		}
		return true;
	}
}